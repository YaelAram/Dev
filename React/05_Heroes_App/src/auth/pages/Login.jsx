import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context/AuthContext';

export function Login() {
    const navigate = useNavigate();
    const { login } = useContext( AuthContext )

    const handleLogin = () => {
        login( 'Yael' );
        navigate( '/heroes/marvel', { replace: true } );
    };

    return (
        <React.Fragment>
            <h1>Login Page</h1>
            <input type="button" value="Login" onClick={ handleLogin } />
        </React.Fragment>
    );
};
