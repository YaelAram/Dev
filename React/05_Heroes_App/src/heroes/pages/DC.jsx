import React from 'react';

import { HeroList } from '../components/HeroList'

export function DC() {
    return (
        <section>
            <h1>DC Comics</h1>
            <HeroList publisher="DC Comics" />
        </section>
    );
};
