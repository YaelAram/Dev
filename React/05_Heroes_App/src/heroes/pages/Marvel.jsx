import React from 'react';
import { HeroList } from '../components/HeroList';

export function Marvel() {
    return (
        <section>
            <h1>Marvel Comics</h1>
            <HeroList publisher="Marvel Comics" />
        </section>
    );
};
