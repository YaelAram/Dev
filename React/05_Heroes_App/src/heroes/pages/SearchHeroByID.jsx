import React, { useMemo } from 'react';
import { useParams, Navigate, useNavigate } from 'react-router-dom';

import { getHeroesByID } from '../helpers'

export function SearchHeroById() {
    const { id } = useParams();
    const hero = useMemo( () => getHeroesByID( id ), [ id ] );

    if( !hero ) return <Navigate to="/heroes/" />

    const { superhero, publisher, alter_ego, first_appearance, characters } = hero;
    const heroIMG = `/assets/heroes/${ id }.jpg`;
    const navigate = useNavigate();
    const onNavigateBack = () => navigate( `/heroes/${ publisher.split( ' ' ).shift() }` )

    return (
        <section>
            <h1>{ superhero }</h1>
            <div className="image-info">
                <img src={ heroIMG } alt={ superhero } className="hero-img" />
                <article className="hero-info">
                    <p><span className="label-hero-info">Publisher: </span>{ publisher }</p>
                    <p><span className="label-hero-info">Alter Ego: </span>{ alter_ego }</p>
                    <p><span className="label-hero-info">First Apperance: </span>{ first_appearance }</p>
                    { 
                        ( characters !== alter_ego ) ?
                        <p><span className="label-hero-info">Characters: </span>{ characters }</p> :
                        undefined
                    }
                    <input type="button" value="Back" onClick={ onNavigateBack } className="primary-button" />
                </article>
            </div>
        </section>
    );
};
