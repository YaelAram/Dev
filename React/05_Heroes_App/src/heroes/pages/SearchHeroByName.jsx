import React, { useMemo, useCallback } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';

import { Alert, SearchBar } from '../../ui';
import { HeroItem } from '../components/HeroItem';
import { getHeroesByName } from '../helpers';

export function SearchHeroByName() {
    const navigate = useNavigate();
    const location = useLocation(); 

    const handleSearch = useCallback( ( searchText ) => navigate( `?q=${ searchText }` ) );
    const heroSearch = useMemo( () => ( new URLSearchParams( location.search ).get( 'q' ) ?? '' ), [ location ] );

    const heroes = useMemo( () => getHeroesByName( heroSearch ), [ heroSearch ] );

    return (
        <section>
            <h1>Search Hero</h1>
            <SearchBar handleSearch={ handleSearch } init={ heroSearch }/>
            {
                ( !heroSearch ) ? undefined :
                    ( heroes.length ) ? 
                        ( 
                            <div className="hero-list">
                                { heroes.map( ( hero ) => <HeroItem key={ hero.id } { ...hero } /> ) }
                            </div>
                        ) :
                        <Alert name={ heroSearch } />
            }
        </section>
    );
};
