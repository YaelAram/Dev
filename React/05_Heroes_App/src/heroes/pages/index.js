export * from './DC';
export * from './Marvel';
export * from './SearchHeroByID'
export * from './SearchHeroByName'
