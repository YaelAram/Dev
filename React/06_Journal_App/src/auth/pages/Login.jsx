import React, { useMemo } from 'react';

import { useDispatch, useSelector } from 'react-redux';

import { Link as RouterLink } from 'react-router-dom';

import { Google } from '@mui/icons-material';
import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material';

import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks';
import { validateEmail, validatePassword } from '../../helpers/validate';
import { startCustomLogin, startGoogleSignIn } from '../../store';

const formData = {
    email: '',
    password: ''
};

const formValidators = {
    email: [ validateEmail, 'Required field, please check you typed a valid email' ],
    password: [ validatePassword, 'Required field, password must have at least 8 characters' ]
};

export function Login() {
    const { status, error } = useSelector( ( state ) => state.auth );
    const dispatch = useDispatch();

    const isChecking = useMemo( () => status === 'checking', [ status ] );

    const { email, password, formValid, emailValid, passwordValid, onInputChange, onReset } = useForm( formData, formValidators );

    const handleSubmit = ( event ) => {
        event.preventDefault();
        dispatch( startCustomLogin( email, password ) );
    };

    const handleGoogleSignIn = () => {
        dispatch( startGoogleSignIn() );
    };

    return (
        <AuthLayout title="Login">
            <form onSubmit={ handleSubmit } className="animate__animated animate__fadeIn animate__faster">
                <Grid container>
                    <Grid item xs={ 12 } sx={ { mt: 2 } }>
                        <TextField label="Email" type="email" placeholder="email@google.com" fullWidth 
                        name="email" value={ email } onChange={ onInputChange }
                        error={ !!emailValid } helperText={ emailValid } />
                    </Grid>

                    <Grid item xs={ 12 } sx={ { mt: 3 } }>
                        <TextField label="Password" type="password" placeholder="password" fullWidth
                        name="password" value={ password } onChange={ onInputChange }
                        error={ !!passwordValid } helperText={ passwordValid } />
                    </Grid>

                    <Grid container spacing={ 2 } sx={ { mb: 2, mt: 1 } }>
                        <Grid item xs={ 12 } display={ error ? '' : 'none' }>
                            <Alert severity="error">{ error }</Alert>
                        </Grid>

                        <Grid item xs={ 12 } sm={ 6 }>
                            <Button type="submit" variant="contained" disabled={ isChecking || !formValid } fullWidth>
                                Login
                            </Button>
                        </Grid>

                        <Grid item xs={ 12 } sm={ 6 }>
                            <Button variant="contained" disabled={ isChecking } fullWidth onClick={ handleGoogleSignIn }>
                                <Google /><Typography sx={ { ml: 1 } }>Google</Typography>
                            </Button>
                        </Grid>
                    </Grid>

                    <Grid container direction="row" justifyContent="end">
                        <Link component={ RouterLink } color="inherit" to="/auth/register">
                            Create new account
                        </Link>
                    </Grid>
                </Grid>
            </form>
        </AuthLayout>
    );
};
