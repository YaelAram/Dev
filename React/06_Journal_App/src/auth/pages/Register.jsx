import React from 'react';

import { Alert, Button, Grid, Link, TextField } from '@mui/material';

import { Link as RouterLink } from 'react-router-dom';
import { AuthLayout } from '../layout/AuthLayout';

import { useForm } from '../../hooks';
import { validateAlpha, validateEmail, validatePassword } from '../../helpers/validate';
import { useDispatch, useSelector } from 'react-redux';
import { startCustomSignIn } from '../../store/auth';
import { useMemo } from 'react';

const formData = {
    displayName: '',
    email: '',
    password: '',
};

const formValidators = {
    displayName: [ validateAlpha, 'Required field, please use only alphabetical characters' ],
    email: [ validateEmail, 'Required field, please check you typed a valid email' ],
    password: [ validatePassword, 'Required field, password must have at least 8 characters' ]
};

export function Register() {
    const dispatch = useDispatch();
    const { status, error } = useSelector( state => state.auth );
    const isChecking = useMemo( () => status === 'checking', [ status ] );
    const { displayName, email, password, form, formValid,
        displayNameValid, emailValid, passwordValid, onInputChange, onReset } = useForm( formData, formValidators );

    const handleSubmit = ( evt ) => {
        evt.preventDefault();
        if( !formValid ) return;
        dispatch( startCustomSignIn( form ) );
    };

    return (
        <AuthLayout title="Create User">
            <form onSubmit={ handleSubmit } className="animate__animated animate__fadeIn animate__faster">
                <Grid container>
                    <Grid item xs={ 12 } sx={ { mt: 2 } }>
                        <TextField label="Name" type="text" placeholder="Name" fullWidth
                        value={ displayName } onChange={ onInputChange } name="displayName"
                        error={ !!displayNameValid } helperText={ displayNameValid } />
                    </Grid>

                    <Grid item xs={ 12 } sx={ { mt: 2 } }>
                        <TextField label="Email" type="email" placeholder="correo@google.com" fullWidth
                        value={ email } onChange={ onInputChange } name="email"
                        error={ !!emailValid } helperText={ emailValid } />
                    </Grid>

                    <Grid item xs={ 12 } sx={ { mt: 3 } }>
                        <TextField label="Password" type="password" placeholder="password" fullWidth
                        value={ password } onChange={ onInputChange } name="password"
                        error={ !!passwordValid } helperText={ passwordValid } />
                    </Grid>

                    <Grid container spacing={ 2 } sx={ { mb: 2, mt: 1 } }>
                        <Grid item xs={ 12 } display={ !!error ? '' : 'none' }>
                            <Alert severity="error">{ error }</Alert>
                        </Grid>
                        <Grid item xs={ 12 }>
                            <Button variant="contained" type="submit" disabled={ !formValid || isChecking } fullWidth>
                                Create Account
                            </Button>
                        </Grid>
                    </Grid>

                    <Grid container direction="row" justifyContent="end">
                        <Link component={ RouterLink } color="inherit" to="/auth/login">
                            I already have an account
                        </Link>
                    </Grid>
                </Grid>
            </form>
        </AuthLayout>
    );
};
