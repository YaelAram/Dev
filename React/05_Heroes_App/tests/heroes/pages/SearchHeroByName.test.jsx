import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchHeroByName } from '../../../src/heroes/pages/SearchHeroByName';

const mockedUseNavigate = jest.fn();

jest.mock( 'react-router-dom', () => ( {
    ...jest.requireActual( 'react-router-dom' ),
    useNavigate: () => mockedUseNavigate
} ) );

describe( 'SearchHeroByName component', () => {

    beforeAll( () => jest.clearAllMocks() );

    test( 'Should show the default page (no query params)', () => {
        const { container } = render(
            <MemoryRouter>
                <SearchHeroByName />
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();
    } );

    test( 'Should show batman in the input text', () => {
        const { container } = render(
            <MemoryRouter initialEntries={ [ '/heroes/search?q=batman' ] }>
                <SearchHeroByName />
            </MemoryRouter>
        );

        expect( screen.getByRole( 'textbox' ).value ).toBe( 'batman' );
        expect( screen.getByRole( 'img' ).src ).toContain( 'batman' );
    } );

    test( 'Should show batman as a query parameter', () => {
        const { container } = render(
            <MemoryRouter initialEntries={ [ '/heroes/search' ] }>
                <SearchHeroByName />
            </MemoryRouter>
        );

        fireEvent.change( screen.getByRole( 'textbox' ), { target: { value: 'batman' } } );
        fireEvent.submit( screen.getByRole( 'form' ) );

        expect( mockedUseNavigate ).toBeCalled();
        expect( mockedUseNavigate ).toBeCalledTimes( 1 );
        expect( mockedUseNavigate ).toBeCalledWith( '?q=batman' );
    } );
} );
