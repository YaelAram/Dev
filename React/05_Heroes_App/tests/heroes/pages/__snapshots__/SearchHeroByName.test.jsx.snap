// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchHeroByName component Should show the default page (no query params) 1`] = `
<div>
  <section>
    <h1>
      Search Hero
    </h1>
    <form
      aria-label="form"
    >
      <input
        class="input-text"
        placeholder="Spider Man"
        type="text"
        value=""
      />
      <input
        class="primary-button"
        type="submit"
        value="Search"
      />
    </form>
  </section>
</div>
`;
